File: examples_C/gauss.cpp
#include <cstdio>
#include <cstring>
#include "origins.hpp"

std::atomic<uint64_t> BaseOriginVector::atomic_id_counts = 0;

static const int n = 4;
void
print_np_mat_ChNbr(OriginDouble a[n][n+1]) {
  printf("INDEX\n");
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n+1; ++j)
      printf("%5ld ", a[i][j].origins[0].symbol_id);
    printf("\n");
  }
  printf("VALUE\n");
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n+1; ++j)
      printf("%e\t", a[i][j].value);
    printf("\n");
  }
}

int main() {
  double b[n][n+1] = {{ 1/1.,1/2.,1/3.,1/4., 10. },
                      { 1/2.,1/3.,1/4.,1/5., 20. },
                      { 1/3.,1/4.,1/5.,1/6., 30. },
                      { 1/4.*(1+0.001),1/5.,1/6.,1/7., 40. }};

  OriginDouble a[n][n+1];
  memset(a, 0, sizeof(a));
  OriginDouble x[n];
  memset(x, 0, sizeof(x));

  // Reading augmented matrix coefficients
  for (int i = 0; i < n; ++i)
    for (int j = 0; j < n+1; ++j)
       a[i][j] = OriginDouble(b[i][j], true);

  printf("Initial System");
  print_np_mat_ChNbr(a);
    
  // Applying Gauss Elimination
  for (int i = 0; i < n; ++i) {
    if (a[i][i] == 0.0)
      exit(1);
    for (int j = i+1; j < n; ++j) {
      OriginDouble ratio = a[j][i]/a[i][i];
      for (int k = 0; k < n+1; ++k)
        a[j][k] = a[j][k] - ratio * a[i][k];
    }
  }

  // Back Substitution
  x[n-1] = a[n-1][n]/a[n-1][n-1];

  for (int i = n-2; i >= 0; --i) {
    x[i] = a[i][n];
    for (int j = i+1; j < n; ++j)
       x[i] = x[i] - a[i][j]*x[j];
    x[i] = x[i]/a[i][i];
  }

  // Displaying solution
  printf("\nRequired solution is: \n");
  for (int i = 0; i < n; ++i) {
    printf("X%d value: %e\n", i, x[i].value);
    printf("Idx\n");
    for (int origin_index = 0; origin_index < x[i].origins_size; ++origin_index)
      printf("[%ld, %e]\n", x[i].origins[origin_index].symbol_id, x[i].origins[origin_index].coefficient*pow(2.0, 53));
    printf("\n");
  }
}

File: examples/example_py/linalg/gauss.py
# Importing NumPy Library
import sys
sys.path.append('../../..')
sys.path.append('/home/ddefour/interflop/chromatic/include')

import numpy as np

from io import StringIO             # To import matrix in martrixmarket format
from scipy.io import mmread


import tracemalloc
import linecache
import time
import os
from collections import Counter

from random import *
from chnbr import ChNbr

def display_top(snapshot, key_type='lineno', limit=3):
    snapshot = snapshot.filter_traces((
        tracemalloc.Filter(False, "<frozen importlib._bootstrap>"),
        tracemalloc.Filter(False, "<unknown>"),
    ))
    top_stats = snapshot.statistics(key_type)

    print("Top %s lines" % limit)
    for index, stat in enumerate(top_stats[:limit], 1):
        frame = stat.traceback[0]
        # replace "/path/to/module/file.py" with "module/file.py"
        filename = os.sep.join(frame.filename.split(os.sep)[-2:])
        print("#%s: %s:%s: %.1f KiB"
              % (index, filename, frame.lineno, stat.size / 1024))
        line = linecache.getline(frame.filename, frame.lineno).strip()
        if line:
            print('    %s' % line)

    other = top_stats[limit:]
    if other:
        size = sum(stat.size for stat in other)
        print("%s other: %.1f KiB" % (len(other), size / 1024))
    total = sum(stat.size for stat in top_stats)
    print("Total allocated size: %.1f KiB" % (total / 1024))

def Gauss( a, x , n):
    # Applying Gauss Elimination
    for i in range(n):
        print("GE(i): ",i)
        if a[i][i] == 0.0:
            sys.exit('Divide by zero detected!')
            
        for j in range(i+1, n):
            ratio = a[j][i]/a[i][i]
            
            for k in range(i+1,n+1):
                a[j][k] = a[j][k] - ratio * a[i][k]

            a[j][i]=0

    # Back Substitution
    x[n-1] = a[n-1][n]/a[n-1][n-1]

    for i in range(n-2,-1,-1):
        x[i] = a[i][n]
        
        for j in range(i+1,n):
            x[i] = x[i] - a[i][j]*x[j]
        
        x[i] = x[i]/a[i][i]

tracemalloc.start()




# Reading number of unknowns
#n = int(input('Enter number of unknowns: '))
# EXEMPLE 1: of a 2x2 Ill conditionned system 
# n = 2
# b = np.array([[1.,1.,2.],[1.,1.001,2]])

#m = mmread('bcspwr01.mtx.gz')
m = mmread('/home/ddefour/interflop/chromatic/examples/example_py/linalg/bcsstk14.mtx.gz')
if (m.A.shape[0] == m.A.shape[1]):
    n = m.A.shape[1]
    print("Handling matrix of shape :", m.A.shape)
else:
    print("Not square matrix !")
    exit()

a = np.array(m.A)
b = np.array([i for i in range(n)])


#start = time.perf_counter()
#x = np.linalg.solve(a, b)
#end = time.perf_counter()
#print(f"\n#(Linalg) Time taken is {end - start}")



# Making numpy array of n x n+1 size and initializing 
# to zero for storing augmented matrix
a_ch = np.zeros((n,n+1), dtype=np.dtype(ChNbr))

# Making numpy array of n size and initializing 
# to zero for storing solution vector
x_ch = np.zeros(n, dtype=np.dtype(ChNbr))

# Reading augmented matrix coefficients
for i in range(n):
    for j in range(n):
        if a[i][j] != 0:
            a_ch[i][j] = ChNbr(a[i][j],follow=True)
        else:
            a_ch[i][j] = 0.0
    # last column B vector
    a_ch[i][n]=ChNbr(float(i+1),follow=True)
    
for i in range(n):
    x_ch[i] = ChNbr(i,follow=True)

#print("Initial System")
#print_np_mat_ChNbr(a)
 

start = time.perf_counter()
Gauss( a_ch, x_ch, n )
end = time.perf_counter()

snapshot = tracemalloc.take_snapshot()
display_top(snapshot)

print("Time taken is ", (end - start))

